agent/consul/gateways/controller_gateways.go

```diff
@@ -2,27 +2,208 @@ package gateways
 
 import (
 	"context"
-
-	"github.com/hashicorp/go-hclog"
-
+	"fmt"
 	"github.com/hashicorp/consul/agent/consul/controller"
 	"github.com/hashicorp/consul/agent/consul/fsm"
+	"github.com/hashicorp/consul/agent/consul/gateways/datastore"
 	"github.com/hashicorp/consul/agent/consul/state"
 	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
 )
 
 type apiGatewayReconciler struct {
-	fsm    *fsm.FSM
 	logger hclog.Logger
+	store  datastore.DataStore
+}
+
+func (r apiGatewayReconciler) retrieveAllRoutesFromStore() ([]structs.BoundRoute, error) {
+	tcpRoutes, err := r.store.GetConfigEntriesByKind(structs.TCPRoute)
+	if err != nil {
+		return nil, err
+	}
+
+	//TODO not implemented
+	//httpRoutes, err := r.store.GetConfigEntriesByKind(structs.HTTPRoute)
+	//if err != nil {
+	//	return nil, err
+	//}
+
+	routes := []structs.BoundRoute{}
+	for _, r := range tcpRoutes {
+		if r == nil {
+			continue
+		}
+		routes = append(routes, r.(*structs.TCPRouteConfigEntry))
+	}
+	//TODO not implemented
+	//for _, r := range httpRoutes {
+	//	routes = append(routes, r.(*structs.HTTPRouteConfigEntry))
+	//}
+	return routes, nil
 }
 
 func (r apiGatewayReconciler) Reconcile(ctx context.Context, req controller.Request) error {
+	entry, err := r.store.GetConfigEntry(req.Kind, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+
+	if entry == nil {
+		r.logger.Info("cleaning up deleted gateway object", "request", req)
+		if err := r.store.Delete(&structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           req.Name,
+			EnterpriseMeta: *req.Meta,
+		}); err != nil {
+			r.logger.Error("error cleaning up deleted gateway object", err)
+			return err
+		}
+		return nil
+	}
+
+	gatewayEntry := entry.(*structs.APIGatewayConfigEntry)
+	err = gatewayEntry.Validate()
+	if err != nil {
+		r.logger.Debug("persisting gateway status", "gateway", gatewayEntry)
+		if updateErr := r.store.UpdateStatus(gatewayEntry); err != nil {
+			return fmt.Errorf("%v: %v", err, updateErr)
+		}
+		return err
+	}
+
+	var boundGatewayEntry *structs.BoundAPIGatewayConfigEntry
+	boundEntry, err := r.store.GetConfigEntry(structs.BoundAPIGateway, req.Name, req.Meta)
+	if err != nil {
+		return err
+	}
+	if boundEntry == nil {
+		boundGatewayEntry = &structs.BoundAPIGatewayConfigEntry{
+			Kind:           structs.BoundAPIGateway,
+			Name:           gatewayEntry.Name,
+			EnterpriseMeta: gatewayEntry.EnterpriseMeta,
+		}
+	} else {
+		boundGatewayEntry = boundEntry.(*structs.BoundAPIGatewayConfigEntry)
+	}
+
+	r.logger.Debug("started reconciling gateway")
+
+	routes, err := r.retrieveAllRoutesFromStore()
+
+	if err != nil {
+		return err
+	}
+
+	metaGateway := &gatewayMeta{
+		BoundGateway: boundGatewayEntry,
+		Gateway:      gatewayEntry,
+	}
+
+	//TODO is this needed? It seems to be for my tests
+	r.reconcileListeners(metaGateway)
```

Comment:
So, I would potentially merge this and the logic from lines 80-88 in something like an `ensureBoundGateway` method, because from the implementation what this is doing is just ensuring that the bound gateway entry initialized or read previously has the proper listener state assigned to it.